Fix very high accuracies feigning as "perfect" 100% plays when actually below

peppy opened the pull request:
Closes https://github.com/ppy/osu/issues/12036.

Note that this does not match stable behaviour, but I think it makes sense. Also as a note, stable will display `0.01` as `01.00%` while we are showing `1.00%`. I think I prefer the padded display that stable does, so if enough people agree I'd be looking to change to that. Not a big deal either way though.

peppy committed:
Fix very high accuracies feigning as "perfect" 100% plays when actually below

peppy committed:
Add test coverage of accuracy formatting function

smoogipoo commented:
Are you sure that the issue isn't `ScoreProcessor` rounding? Which it shouldn't be doing, btw.

peppy commented:
I'd hope not (that would break a lot more than just the display, I think?). The tests confirm that this was an issue with the display function either way.

peppy committed:
Remove accuracy rounding at a ScoreProcessor level

peppy committed:
Add TODO marking incorrect EF core data type

As mentioned in the comment, we probably don't need to be storing this
in the database in the first place (as it should be able to be
calculated from the other statistics we have available to us). Something
to consider when we refactor the database backend.

peppy committed:
Remove unnecessary using

bdach commented on osu.Game.Tests/NonVisual/FormatUtilsTest.cs:
```diff
@@ -0,0 +1,25 @@
+        [TestCase(0, "0.00%")]
+        [TestCase(0.01, "1.00%")]
+        [TestCase(0.9899, "98.99%")]
+        [TestCase(0.989999, "99.00%")]
```
I'm not sure I agree with this test case. I understand that 100% is a special accuracy, but you could argue that if 99.999% doesn't get rounded up to a full percent, then neither should 98.999%, 89.999%, and so forth. I would probably do something like this instead:

```csharp
            int truncated = (int)(accuracy * 10000);
            return $"{truncated / 10000.0f:0.00%}";
```

It's a bit harsh maybe to just truncate, but in a rhythm game I'd say it's warranted to be stricter, and it's a third decimal place anyway. Of course it gets hazy around the round numbers, but again, I'm not sure that 89 to 90% should differ much from 99 to 100%.

peppy replied:
From the angle that we have other cutoffs for grades which are always round ints, i think you may be making some sense here. Else we were get similar complaints "I got 90% but not an A" or the likes.

peppy replied:
Have applied this change, I think it feels better.

bdach commented on osu.Game/Utils/FormatUtils.cs:
```diff
@@ -12,14 +12,14 @@ public static class FormatUtils
-        /// <param name="accuracy">The accuracy to be formatted</param>
-        /// <returns>formatted accuracy in percentage</returns>
-        public static string FormatAccuracy(this decimal accuracy) => $"{accuracy:0.00}%";
+            return $"{accuracy:0.00%}";
```
With the way this method is written, the test cases attached in this PR fail on systems that use a different decimal separator than the full stop due to regional settings (i.e. mine, as Polish uses a decimal comma rather than a full stop).

Probably need to either rewrite the test to format both values according to culture settings, or to put a `FormattableString.Invariant()` call around this, to avoid confusing others as to why those tests are failing.

peppy replied:
Considered this from a CI perspective, but not from other-users-running-tests.

bdach commented:
Also I checked the math using the screenshot posted in the issue, and this would have popped up regardless of the rounding in `ScoreInfo`.

![2021-03-25-204338_513x129_scrot](https://user-images.githubusercontent.com/20418176/112537676-4583f080-8daf-11eb-8835-e491cd908edc.png)

`string.Format()` does spit out `100.00%` by default in such a case.

peppy committed:
Change rounding to use a more general flooring approach

peppy committed:
Add the ability to specify format provider (to make tests culture invariant)

bdach committed:
Merge branch 'master' into fix-accuracy-formatting

bdach commented:
Oh, re: the leading zero thing - I think prefer without, `01.00%` looks a bit... odd, to me. Would that leading zero be for visual alignment only?

peppy commented:
Leading zero change can stay as is for now I think. Can be reconsidered at a later point in time

bdach committed:
Merge branch 'master' into fix-accuracy-formatting
